tensorflow/core/kernels/BUILD

```diff
@@ -3983,7 +4005,11 @@ tf_kernel_library(
         "topk_op_gpu_int8.cu.cc",
         "topk_op_gpu_uint8.cu.cc",
     ],
-    deps = NN_DEPS + if_cuda(["@cub_archive//:cub"]),
+    deps = NN_DEPS + if_cuda([
```

Comment:
The `if_*_is_configured` variants are needed when both the `cuda` and `rocm` need to specify the same dependency.

Take for example the target for the `softmax_op` in this file.
Both `cuda` and `rocm` paths have a dependency on the the `:reduction_ops` target. If we use the `if_cuda` / `if_rocm` variants, then bazel will give the following "duplicate dependency" error

```
...
ERROR: /root/tensorflow/tensorflow/core/kernels/BUILD:3963:1: Label '//tensorflow/core/kernels:reduction_ops' is duplicated in the 'deps' attribute of rule 'softmax_op'
...
```

Using the `if_rocm_is_configured` / `if_cuda_is_confgiured` helps avoid that error.

I do not understand bazel good enough to be able to explain the reason for the difference in behaviour :(

The definitions for the `if_*` can be found here :

https://github.com/tensorflow/tensorflow/blob/master/third_party/gpus/cuda/build_defs.bzl.tpl#L25
https://github.com/tensorflow/tensorflow/blob/master/third_party/gpus/rocm/build_defs.bzl.tpl#L37

Reply:
one guess as to the reason of the difference is that 
* the `if_cuda` / `if_rocm` versions return a `select` which will resolve to either cuda/rocm value or empty *upon evaluation*, 
* where as the `if_cuda_is_configured` / `if_rocm_is_configured` variants will directly return the cuda/rocm value or empty. 

Because of this difference, my guess is that the bazel code that does the dependency analysis will see a duplicate dependency in the `if_cuda` / `if_rocm` case (because both will return `select`s which will contain the same dependency...the dependency code has no reason to believe that the two are mutually exclusive)

In the `if_cuda_is_configured` / `if_rocm_is_configured` case, only one of them will return a non-empty value (because you typically compile with either `--config=rocm` or `--config=cuda`, but not both!), and hence the dependency analysis code will see no duplicates.

Reply:
Correct, `if_cuda / if_rocm` returns a `select` statement. When you build with `--config=rocm / cuda`,  blaze will pick the `if_true` macro parameter, otherwise the `if_false` macro parameter. If you do `if_cuda([], if_false=":x") + if_rocm([], if_false=":x")`, without `--config=rocm / cuda` you will get `:x` twice.

You can't nest `select` statements. So 'if_cuda([], if_false = if_rocm([], if_false = ":x")` is not a valid work-around. But we could easily write a `if_gpu(if_cuda, if_rocm, otherwise)` macro. Would that make things cleaner?

The `if_rocm/cuda_is_configured` macros are *not* affected by `--config=rocm / cuda`. The implementation of those macros is generated as part of the repository rule (`local_config_cuda/rocm`) and depends on the `TF_NEED_CUDA/ROCM` environment variable. Those environment variables are usually written to .tf_configure.bazelrc during the configuration process. That process also enables building with `--config=rocm / cuda` by default (written to .tf_configure.bazelrc as well). I'm pretty sure we are not testing building with `--action_env=TF_NEED_CUDA=1` but without `--config=cuda` regularly, so it might be (likely is) broken. I understand that this makes it difficult to pick the right macro (`if_rocm` vs `if_rocm_is_configured`), but hopefully understanding the intention will allow you to do better than trial-and-error (or use `if_rocm_is_configured` everywhere).

Reply:
>Correct, if_cuda / if_rocm returns a select statement. When you build with --config=rocm / cuda, blaze will pick the if_true macro parameter, otherwise the if_false macro parameter. If you do if_cuda([], if_false=":x") + if_rocm([], if_false=":x"), without --config=rocm / cuda you will get :x twice.

understood. though in this case the common dependency is in the `if_true` part for both `cuda` and `rocm`. My thought process was that the bazel does not assume the two (`rocm`/`cuda`)  are mutually exclusive, and hence complains about duplicate dependencies for the case when both take the true branch.

>But we could easily write a if_gpu(if_cuda, if_rocm, otherwise)` macro. Would that make things cleaner?

maybe.  @whchung what do you think about the `if_gpu (if_cuda_arg, if_rocm_arg, default_arg)` suggestion?

> but hopefully understanding the intention will allow you to do better than trial-and-error

better understanding is always helpful. thanks again for taking the time out to explain.

Reply:
Deven, you are correct, and I'm surprised ;-)
The following indeed produces 'duplicate main.cc in srcs', even without any '--define':

```python
config_setting(
      name = "foo",
      values = {
          "define": "foo=1"
      }
)

config_setting(
      name = "bar",
      values = {
          "define": "bar=1"
      }
)

cc_binary(
    name = "main",
    srcs = select({
        ":foo": ["main.cc"],
        "//conditions:default": []
    }) + select({
        ":bar": ["main.cc"],
        "//conditions:default": []
    }),
)
```
The same with a single select works in all cases (well, except if you define both foo=1 and bar=1):
```python
    srcs = select({
        ":foo": ["main.cc"],
        ":bar": ["main.cc"],
        "//conditions:default": ["main.cc"]
    }),
```